### ISSUE TITLE
Dealing with unstable to stable

### ISSUE THREAD

**Issue by ThePotatoGuy** [Monika-After-Story/MonikaModDev#8086]
reverting back from unstable used to be supported, but after #6913, this is no longer the case.

Since update scripts run in unstable versions, the potential for data corruption is much higher for unstable users, and the only safe window for reverting to stable is when unstable and stable are on the same version, which is something that cannot be easily determined by non-technical users.

So we need to think of ways to address this. Here is a list of the primary issues that need to be addressed:
1. Reverting to an unstable using the checkbox should be either limited to non-breaking situations or made possible to do without breaking the game.
2. Running a stable version with an unstable persistent should not cause game-breaking issues or data corruption.

Here's some potential solutions I've thought of:
1. Lock down unstable checkbox when a non-revertable version is loaded.
    * HOW - we can programmatically determine a revertable version by checking the version numbers. Non-revertable unstables will have an extra version number. 
    * GOOD - prevents reverting from unstable to stable
    * BAD - However, this would mean a broken unstable build will screw users until we can push fixes out. Additionally, with our longer main release cycle, reverting can only happen every few months or so.
    * Issue 1 would be handled neatly.
    * Issue 2 would **not** be handled.
2. Add scripts for reverting update scripts.
    * HOW - Would have to add some additional framework to run reverse update scripts. 
    * GOOD - unstables can be safely reverted from via the checkbox - a bit better for users rather then forcing them so stay with unstable for a bit
    * BAD - This goes against part of the spirit of update scripts, which are designed as one-way scripts. Also means additional code to test and maintain per unstable. 
    * Issue 1 would be handled, with some effort per unstable.
    * Issue 2 would **not** be handled, since an unstable reverse script would not be available in a stable build.
3. Prohibit unstable update scripts that will result in a persistent that cannot run in unstable
    * HOW - basically a procedural change to how we do update scripts in unstable. This would be somewhat a reversal back to unstable practices before #6913, where unstable features that relied on persistent changes could not be tested. 
    * GOOD - handles both issues 
    * BAD - I don't know exactly what kinds of unstable updates scripts would work without causing persistent issues in stable, so I don't know if this really means unstable update scripts become useless.
    * Issue 1 would be handled, since reversion should be possible again.
    * Issue 2 would be handled, since the persistent won't be broken for stables.
4. Sandbox unstable changes with an automatic system for restoring data on return to stable
    * HOW - when we do an unstable script, we will need to mark (or manually save) the variables that we change to a separate persistent var. think of this var as a log of the changes. When a stable build is loaded, we can check if the previous version was an unstable of a higher version number, and if so, pull the persistent var and revert the marked variables to their previous values. This process would have to happen earlier than traditional update scripts, but after persistent backups are made - so likely in the 99x init range. 
    * GOOD - handles both issues
    * BAD - requires an entirely new framework to be build. Also needs to be in a stable version *before* regular unstables can be released, so unstable update scripts would have to be postponed (or limited to non-breaking) until the first stable release the framework is released. Additional code and testing per unstable version would also be required.
    * Issue 1 would be handled, since reversion should be possible again.
    * Issue 2 would be handled, since the persistent would be reverted on stables.
5. Solution 1 + don't load persistents that have unstable version numbers in stables
    * HOW - apply the version number logic from solution 1 to the load process for persistents. This would to be something before/during the persistent backup process. We would need to either try to load a persistent that has a safe version OR hardstop with a chibikia screen mentioning the incompatibilities. Also, making an extra persistent backup when changing from a stable to an unstable version might be a good way to almost guarantee a safe persistent can be loaded.
    * GOOD - handles both issues
    * BAD - need some additional -999 init handling (which is already somewhat crowded and dangerous), but no other major issues
    * Issue 1 would be handled via solution 1
    * Issue 2 would be handled by persistent checking early in init.
6. Solution 1 + part of Solution 5 + force update to unstable.
    * HOW - same as solution 1 for checking versions. Loading will have to check persistents, but we don't do backup checking nor attempt to load a stable persistent. Instead, the current persistent will be backed up as special unstable persistent, a fresh persistent is forced, and splash will force the user to update to unstable (probably via some chibika dialogue then the updater screen). When game is relaunched after the update, the special unstable persistent backup is loaded as the persistent. (for all nuances, see my comment here )
    * GOOD - handles both issues, while also making it possible to get an unstable build without actually having the game run normally 
    * BAD - see solution 5. I can't think of any reasons this would be bad otherwise.

If you have any additional ideas, please comment below. I'll add more if I think of anything else.

**Comment by ThePotatoGuy** [comment]
> removing the unstable option in settings altogether.
Definitely no plans to remove the unstable checkbox completely, the ideal solution is something like Steam's opt-in betas, but we don't have that luxury so an in-game toggle will suffice.

Anyway, I've been working on this for a bit, and I've come up with another idea (solution 6) that should address some of the concerns of not having a publicly available download link for unstables. 

Solution 6 is similar to solution 5, but instead of trying hard to load a valid, stable persistent, chibika will just be invoked to force the user to update to the unstable version. The full workflow is as follows:
1. Incompatible persistent detected on load
2. the incompat persistent is renamed to a special name like `persistent_unstable`. Since the persistent is renamed, the game generates a new fresh persistent.
    * if renaming fails, (aka making a backup fails), then an exception is raised and we hardstop. (avoids destroying the unstable persistent on load)
3. Splash branches to some chibiki dialogue explaining the situation. This branch eventually forces the updater to run in unstable mode (updating to unstable).
4. Update completes, game is restarted.
5. the special persistent is detected on load, and copied back as `persistent` if its compatible with the installed build. 
6. Game continues loading as normal

### nuances with step 5
Couple of paths if a special persistent is detected on load:
* If the special persistent is corrupt, it will be deleted and game behavior will continue normally (assuming the regular persistent is fine as well)
* if the game is stable and the special persistent is incompatible, the branching behavior in step 3 will continue to happen until either the game is updated to unstable/new stable, or the special persistent backup is deleted/no longer detected. This would likely happen if the unstable update fails to complete.
* if the special persistent fails to be copied back, then an exception is raised with hardstop. No chibika branch because the existence of the special persistent drives some of this behavior, so the file must be deleted for things to work well.

**Comment by Booplicate** [comment]
> 2. the incompat persistent is moved to a special name like `persistent_unstable`. Since the persistent is moved, the game generates a new fresh persistent.
>    
>    * if renaming fails, (aka making a backup fails), then an exception is raised and we hardstop. (avoids destroying the unstable persistent on load)

> 5. the special persistent is detected on load, and copied back as `persistent` if its compatible with the installed build.

Do you think renaming persistent and generating a fresh one is mandatory? I may be missing something, but if we removed that part, we wouldn't lose much, yet would avoid potential permission issues and would do less writing/copying (less chance for corruption). Plus, we wouldn't leave people with a fresh persistent if something goes wrong.

**Comment by ThePotatoGuy** [comment]
going fresh is mandatory since we could crash in init if stable can't handle unstable persistent data.

**Comment by ThePotatoGuy** [comment]
[closing as implemented]
